### PR TITLE
C++: Introduce wrapper for the semantic range analysis mimicking the simple one

### PR DESCRIPTION
**Pull request by jketema**

### CODE REVIEW

File: cpp/ql/lib/experimental/semmle/code/cpp/semantic/analysis/SimpleRangeAnalysis.qll
Comment (jketema) [inline cpp/ql/lib/experimental/semmle/code/cpp/semantic/analysis/SimpleRangeAnalysis.qll:27]: The QLDoc here was directly copied from the simple range analysis library. What I'm not convinced about is whether we satisfy the part of the QLDoc that talks about casts, or whether some additional logic is needed for that.

File: cpp/ql/lib/experimental/semmle/code/cpp/semantic/analysis/SimpleRangeAnalysis.qll
Comment (jketema) [inline cpp/ql/lib/experimental/semmle/code/cpp/semantic/analysis/SimpleRangeAnalysis.qll:71]: Note that the simple range analysis implementation is more complex:
```codeql
    analyzableExpr(expr) and
    (
      not exists(lowerBound(expr)) or
      not exists(upperBound(expr)) or
      lowerBound(expr) > upperBound(expr)
    )
```
Do we need something similar here?

File: cpp/ql/lib/experimental/semmle/code/cpp/semantic/analysis/SimpleRangeAnalysis.qll
Comment (MathiasVP) [reply cpp/ql/lib/experimental/semmle/code/cpp/semantic/analysis/SimpleRangeAnalysis.qll:27]: I think the above implementation handles casts correctly. If `expr` is a conversion then `i.getAst() = expr` will be satisfied for some `ConvertInstruction`, and we'll compute the bound for the result of the conversion.

And if `expr` is not a conversion then `i.getAst() = expr` will be satisfied by some non-`ConvertInstruction`, and we'll compute the bound for the unconverted expression.

Now, I don't think we actually properly compute any useful bounds for `ConvertInstruction`s. But hopefully this will change soon 🤞.

File: cpp/ql/lib/experimental/semmle/code/cpp/semantic/analysis/SimpleRangeAnalysis.qll
Comment (MathiasVP) [reply cpp/ql/lib/experimental/semmle/code/cpp/semantic/analysis/SimpleRangeAnalysis.qll:71]: I _think_ (but I'm not sure) that all the `analyzableExpr` part was to exclude expressions for which we could only infer useless bounds based on the lower/upper bounds of their type, and we wanted to exclude those from this predicate since. I'm not quite sure why they wouldn't be excluded by the `lowerBound(expr) > upperBound(expr)` conjunct, though 🤔.

File: cpp/ql/lib/experimental/semmle/code/cpp/semantic/analysis/SimpleRangeAnalysis.qll
Comment (MathiasVP) [inline cpp/ql/lib/experimental/semmle/code/cpp/semantic/analysis/SimpleRangeAnalysis.qll:98]: The implementations of these can properly be copy/pasted from `SimpleRangeAnalysis`:
```ql
predicate convertedExprMightOverflowNegatively(Expr expr) {
  exprMightOverflowNegatively(expr) or
  convertedExprMightOverflowNegatively(expr.getConversion())
}
```

File: cpp/ql/lib/experimental/semmle/code/cpp/semantic/analysis/SimpleRangeAnalysis.qll
Comment (MathiasVP) [inline cpp/ql/lib/experimental/semmle/code/cpp/semantic/analysis/SimpleRangeAnalysis.qll:113]: Same with this one.

File: cpp/ql/lib/experimental/semmle/code/cpp/semantic/analysis/SimpleRangeAnalysis.qll
Comment (MathiasVP) [inline cpp/ql/lib/experimental/semmle/code/cpp/semantic/analysis/SimpleRangeAnalysis.qll:120]: Same with this one.

**Review by MathiasVP — COMMENTED** [review]
Generally, LGTM! There are a couple of the predicates that are just shallow wrappers around some of the other predicates, and we might as well implement those right away.

I'd like @rdmarsh2's eyes on this PR as well, though.

File: cpp/ql/lib/experimental/semmle/code/cpp/semantic/analysis/SimpleRangeAnalysis.qll
Comment (jketema) [reply cpp/ql/lib/experimental/semmle/code/cpp/semantic/analysis/SimpleRangeAnalysis.qll]: added

File: cpp/ql/lib/experimental/semmle/code/cpp/semantic/analysis/SimpleRangeAnalysis.qll
Comment (jketema) [reply cpp/ql/lib/experimental/semmle/code/cpp/semantic/analysis/SimpleRangeAnalysis.qll]: added

File: cpp/ql/lib/experimental/semmle/code/cpp/semantic/analysis/SimpleRangeAnalysis.qll
Comment (jketema) [reply cpp/ql/lib/experimental/semmle/code/cpp/semantic/analysis/SimpleRangeAnalysis.qll]: added

**Review by MathiasVP — APPROVED** [review]
LGTM from me. I'll let @rdmarsh2 take a look as well before merging it.

**Review by rdmarsh2 — APPROVED** [review]

**Pull request merged**
